lib/taste_tester/commands.rb

```diff
@@ -102,29 +102,47 @@ def self.test
         TasteTester::Hooks.pre_test(TasteTester::Config.dryrun, repo, hosts)
       end
       tested_hosts = []
+      host_threads = []
+      connect_failures = 0
       hosts.each do |hostname|
-        host = TasteTester::Host.new(hostname, server)
+        # Poor man thread pool manager: keeping it simple
+        nb_threads_over_max = host_threads.length - TasteTester::Config.parallel_hosts
+        if nb_threads_over_max >= 0
+          host_threads[nb_threads_over_max].join
+        end
+        host_threads << Thread.new do
+          Thread.current[:hostname] = hostname
+          Thread.current.report_on_exception = false
+          TasteTester::Host.new(hostname, server).test
+        end
+      end
+      host_threads.each do |host_thread|
         begin
-          host.test
+          hostname = host_thread[:hostname]
+          host_thread.join
           tested_hosts << hostname
         rescue TasteTester::Exceptions::AlreadyTestingError => e
           logger.error("User #{e.username} is already testing on #{hostname}")
+        rescue TasteTester::Exceptions::SshError
+          logger.error("Cannot connect to #{hostname}")
+          connect_failures += 1
         end
       end
       unless TasteTester::Config.skip_post_test_hook ||
           TasteTester::Config.linkonly
         TasteTester::Hooks.post_test(TasteTester::Config.dryrun, repo,
                                      tested_hosts)
       end
-      # Strictly: hosts and tested_hosts should be sets to eliminate variance in
-      # order or duplicates. The exact comparison works here because we're
-      # building tested_hosts from hosts directly.
-      if tested_hosts == hosts
+      if tested_hosts.to_set == hosts.to_set
         # No exceptions, complete success: every host listed is now configured
         # to use our chef-zero instance.
         exit(0)
       end
       if tested_hosts.empty?
+        if connect_failures > 0
+          # All the hosts we had failed, with at least one because of ssh
```

Comment:
yeah but we will lose here what @malmond77 / @joshuamiller01  want - whether he needs to just try new hosts because they're already in testing, or if there was some other systemic issue. This is important for `aftertaste`.

With this new code if 9 hosts are already-tested and one has a read-only filesystem or something, instead of aftertaste just picking new hosts (which it should), it'll think they all failed for systemic issues and fail altogether.

So... this will have very adverse affects on aftertaste, so the three of you should sort that out amongst yourselves.

Reply:
I checked the code, and as of today, `aftertaste` uses Python's asyncio to do this. So basically it runs one `taste-tester` command per host. So this will work, as for one host, you'll get the same exit code.
And honestly, it cannot really change anything, 'cause before it would barf at the first host that failed, so you'd have to check the output or the hosts to figure out which one. Here, if you have one bad hosts at the 5th element of 10, instead of putting the first 4 to testing and crashing returning 1 (default on exception iirc), it will put all 9, and return 2.